tests/v2/validation/observability/stackstate_test.go

```diff
@@ -0,0 +1,271 @@
+	config.LoadConfig(stackStateConfigFileKey, &stackstateConfigs)
+
+	if !initialStackstateAgent.IsAlreadyInstalled {
+		systemProject, err := projects.GetProjectByName(client, ss.cluster.ID, systemProject)
+		require.NoError(ss.T(), err)
+		systemProjectID := strings.Split(systemProject.ID, ":")[1]
+
+		ss.T().Log("Installing stackstate agent chart with the version before the latest version")
+		err = charts.InstallStackstateAgentChart(client, ss.stackstateAgentInstallOptions, stackstateConfigs.ClusterApiKey, stackstateConfigs.Url, systemProjectID)
+		require.NoError(ss.T(), err)
+
+		ss.T().Log("Verifying the deployments of stackstate agent chart to have expected number of available replicas")
+		err = extencharts.WatchAndWaitDeployments(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+		require.NoError(ss.T(), err)
+
+		ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+		err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+		require.NoError(ss.T(), err)
+	}
+
+	stackstateAgentChartPreUpgrade, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+
+	require.NoError(ss.T(), err)
+
+	// Validate current version of stackstate agent is one of the versions before latest
+	chartVersionPreUpgrade := stackstateAgentChartPreUpgrade.ChartDetails.Spec.Chart.Metadata.Version
+	require.Contains(ss.T(), versionsList[1:], chartVersionPreUpgrade)
+
+	ss.stackstateAgentInstallOptions.Version, err = client.Catalog.GetLatestChartVersion(charts.StackstateK8sAgent, rancherPartnerCharts)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Upgrading stackstate agent chart to the latest version")
+	err = charts.UpgradeStackstateAgentChart(client, ss.stackstateAgentInstallOptions, stackstateConfigs.ClusterApiKey, stackstateConfigs.Url)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the deployments of stackstate agent chart to have expected number of available replicas")
+	err = extencharts.WatchAndWaitDeployments(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+	err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+	require.NoError(ss.T(), err)
+
+	stackstateAgentChartPostUpgrade, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	ss.T().Log("Comparing installed and desired stackstate agent versions")
+	chartVersionPostUpgrade := stackstateAgentChartPostUpgrade.ChartDetails.Spec.Chart.Metadata.Version
+	require.Equal(ss.T(), ss.stackstateAgentInstallOptions.Version, chartVersionPostUpgrade)
```

Comment:
Is there a way to add some server-side check too? i.e. isn't there a case that the agent installs fine, but upstream the stackstate-server isn't able to see the agent anymore? This may be a stackstate-standalone test case but, might be worth adding a `future works` ticket.

Reply:
we have a ticket for it already: https://github.com/rancher/qa-tasks/issues/1595 this will take care of validations on the stackstate itself.